File: detector.py
import math
import statistics
from datetime import datetime, timezone
from config import (
    Z_SCORE_THRESHOLD,
    VOLUME_PCT_THRESHOLD,
    PRE_RESOLUTION_HOURS,
    SUSPICION_THRESHOLD,
    ROLLING_WINDOW,
)
from db import get_recent_trade_sizes


def score_trade(trade: dict, market: dict, wallet: dict = None) -> tuple:
    """
    Returns (suspicion_score, [list of signal descriptions]).
    Score > SUSPICION_THRESHOLD means fire an alert.
    wallet: optional dict with first_trade_ts and trade_count from DB cache.
    """
    signals = []
    large_score = _large_trade_score(trade, market, signals)
    timing_multiplier = _timing_multiplier(trade, market, signals)
    wallet_multiplier = _wallet_multiplier(wallet, trade["timestamp"], signals)

    score = large_score * timing_multiplier * wallet_multiplier

    return score, signals


def _large_trade_score(trade: dict, market: dict, signals: list) -> float:
    """
    Returns a score >= 1.0 if the trade is unusually large, else 0.0.
    Checks two sub-signals: Z-score vs rolling history, and % of 24h volume.
    """
    size = trade["size"]
    if size <= 0:
        return 0.0

    # --- Z-score against rolling window ---
    recent_sizes = get_recent_trade_sizes(trade["market_id"], ROLLING_WINDOW)
    z_score_hit = False
    if len(recent_sizes) >= 10:
        mean = statistics.mean(recent_sizes)
        stdev = statistics.stdev(recent_sizes) if len(recent_sizes) > 1 else 0
        if stdev > 0:
            z = (size - mean) / stdev
            if z >= Z_SCORE_THRESHOLD:
                z_score_hit = True
                signals.append(f"Large position ({z:.1f}x sigma above avg size of ${mean:,.0f})")

    # --- % of 24h volume ---
    vol_pct_hit = False
    volume_24h = market.get("volume_24h", 0)
    if volume_24h > 0:
        pct = (size / volume_24h) * 100
        if pct >= VOLUME_PCT_THRESHOLD:
            vol_pct_hit = True
            signals.append(f"Trade is {pct:.1f}% of 24h market volume (${volume_24h:,.0f})")

    if z_score_hit or vol_pct_hit:
        return 1.0
    return 0.0


def _timing_multiplier(trade: dict, market: dict, signals: list) -> float:
    """
    Returns a multiplier based on proximity to market resolution.
    1.0 = no timing signal; higher = closer to end date.
    """
    end_date_str = market.get("end_date", "")
    if not end_date_str:
        return 1.0

    try:
        end_dt = _parse_iso(end_date_str)
    except (ValueError, TypeError):
        return 1.0

    try:
        trade_dt = _parse_iso(trade["timestamp"])
    except (ValueError, TypeError):
        return 1.0

    hours_remaining = (end_dt - trade_dt).total_seconds() / 3600

    if hours_remaining < 0:
        # Market already resolved, skip timing signal
        return 1.0

    if hours_remaining <= PRE_RESOLUTION_HOURS:
        # Score scales from 1.5 (at 48h) up to 3.0 (at 0h)
        multiplier = 1.5 + 1.5 * (1 - hours_remaining / PRE_RESOLUTION_HOURS)
        signals.append(
            f"Pre-resolution: {hours_remaining:.1f}h before market closes"
        )
        return multiplier

    return 1.0


def _wallet_multiplier(wallet: dict, trade_ts: str, signals: list) -> float:
    """
    Boosts score for wallets that are new or have very few trades.
    Returns 1.0 if no wallet data available (neutral — don't penalize missing data).
    """
    if not wallet:
        return 1.0

    trade_count = wallet.get("trade_count") or 0
    first_trade_ts = wallet.get("first_trade_ts")

    boost = 1.0

    # Very inexperienced wallet
    if trade_count < 5:
        boost = max(boost, 2.0)
        signals.append(f"Fresh wallet: only {trade_count} total trades on Polymarket")
    elif trade_count < 20:
        boost = max(boost, 1.5)
        signals.append(f"New wallet: only {trade_count} total trades on Polymarket")

    # Young account by age
    if first_trade_ts:
        try:
            trade_dt = _parse_iso(trade_ts)
            account_age_days = (trade_dt.timestamp() - first_trade_ts) / 86400
            if account_age_days < 7:
                boost = max(boost, 2.0)
                signals.append(f"New account: first trade was {account_age_days:.1f} days ago")
            elif account_age_days < 30:
                boost = max(boost, 1.3)
                signals.append(f"Young account: first trade was {account_age_days:.0f} days ago")
        except (ValueError, TypeError):
            pass

    return boost


def is_suspicious(score: float) -> bool:
    return score >= SUSPICION_THRESHOLD


def _parse_iso(s: str) -> datetime:
    """Parse ISO 8601 / Unix timestamp string to aware UTC datetime."""
    s = str(s).strip()
    # Handle Unix timestamps (seconds or milliseconds)
    if s.isdigit():
        ts = int(s)
        if ts > 1e12:
            ts //= 1000
        return datetime.fromtimestamp(ts, tz=timezone.utc)
    # Handle ISO strings
    s = s.replace("Z", "+00:00")
    return datetime.fromisoformat(s)

File: backend/src/services/detector.ts
import { Market, Trade, Wallet, ScoringResult, SuspicionSignal, DetectionConfig } from '../../../shared/types';
import { getRecentTradeSizes } from '../db';

/**
 * Score a trade for suspiciousness
 * Returns {score, signals} where score > threshold means fire an alert
 */
export function scoreTrade(
  trade: Omit<Trade, 'createdAt'>,
  market: Market,
  wallet: Partial<Wallet> | null,
  config: DetectionConfig
): ScoringResult {
  const signals: SuspicionSignal[] = [];

  const largeScore = getLargeTradeScore(trade, market, signals, config);
  const timingMultiplier = getTimingMultiplier(trade, market, signals, config);
  const walletMultiplier = getWalletMultiplier(wallet, trade.timestamp, signals);

  const score = largeScore * timingMultiplier * walletMultiplier;

  return { score, signals };
}

/**
 * Returns true if the score exceeds the suspicion threshold
 */
export function isSuspicious(score: number, config: DetectionConfig): boolean {
  return score >= config.suspicionThreshold;
}

/**
 * Returns a score >= 1.0 if the trade is unusually large, else 0.0
 * Checks two sub-signals: Z-score vs rolling history, and % of 24h volume
 */
function getLargeTradeScore(
  trade: Omit<Trade, 'createdAt'>,
  market: Market,
  signals: SuspicionSignal[],
  config: DetectionConfig
): number {
  const size = trade.size;
  if (size <= 0) {
    return 0.0;
  }

  let zScoreHit = false;
  let volPctHit = false;

  // --- Z-score against rolling window ---
  const recentSizes = getRecentTradeSizes(trade.marketId, config.rollingWindow);
  if (recentSizes.length >= 10) {
    const mean = recentSizes.reduce((a, b) => a + b, 0) / recentSizes.length;
    const stdev = calculateStdDev(recentSizes, mean);

    if (stdev > 0) {
      const z = (size - mean) / stdev;
      if (z >= config.zScoreThreshold) {
        zScoreHit = true;
        signals.push({
          type: 'large_position',
          description: `Large position (${z.toFixed(1)}x sigma above avg size of $${mean.toFixed(0)})`,
          value: z,
        });
      }
    }
  }

  // --- % of 24h volume ---
  const volume24h = market.volume24h || 0;
  if (volume24h > 0) {
    const pct = (size / volume24h) * 100;
    if (pct >= config.volumePctThreshold) {
      volPctHit = true;
      signals.push({
        type: 'volume_pct',
        description: `Trade is ${pct.toFixed(1)}% of 24h market volume ($${volume24h.toFixed(0)})`,
        value: pct,
      });
    }
  }

  return zScoreHit || volPctHit ? 1.0 : 0.0;
}

/**
 * Returns a multiplier based on proximity to market resolution
 * 1.0 = no timing signal; higher = closer to end date
 */
function getTimingMultiplier(
  trade: Omit<Trade, 'createdAt'>,
  market: Market,
  signals: SuspicionSignal[],
  config: DetectionConfig
): number {
  if (!market.endDate) {
    return 1.0;
  }

  try {
    const endDate = parseIsoOrUnix(market.endDate);
    const tradeDate = parseIsoOrUnix(trade.timestamp);

    const hoursRemaining = (endDate.getTime() - tradeDate.getTime()) / (1000 * 60 * 60);

    if (hoursRemaining < 0) {
      // Market already resolved, skip timing signal
      return 1.0;
    }

    if (hoursRemaining <= config.preResolutionHours) {
      // Score scales from 1.5 (at 48h) up to 3.0 (at 0h)
      const multiplier = 1.5 + 1.5 * (1 - hoursRemaining / config.preResolutionHours);
      signals.push({
        type: 'pre_resolution',
        description: `Pre-resolution: ${hoursRemaining.toFixed(1)}h before market closes`,
        value: hoursRemaining,
      });
      return multiplier;
    }
  } catch (error) {
    // Invalid date format, return neutral
    return 1.0;
  }

  return 1.0;
}

/**
 * Boosts score for wallets that are new or have very few trades
 * Returns 1.0 if no wallet data available (neutral — don't penalize missing data)
 */
function getWalletMultiplier(
  wallet: Partial<Wallet> | null,
  tradeTimestamp: string,
  signals: SuspicionSignal[]
): number {
  if (!wallet) {
    return 1.0;
  }

  const tradeCount = wallet.tradeCount || 0;
  const firstTradeTs = wallet.firstTradeTs;

  let boost = 1.0;

  // Very inexperienced wallet
  if (tradeCount < 5) {
    boost = Math.max(boost, 2.0);
    signals.push({
      type: 'fresh_wallet',
      description: `Fresh wallet: only ${tradeCount} total trades on Polymarket`,
      value: tradeCount,
    });
  } else if (tradeCount < 20) {
    boost = Math.max(boost, 1.5);
    signals.push({
      type: 'new_wallet',
      description: `New wallet: only ${tradeCount} total trades on Polymarket`,
      value: tradeCount,
    });
  }

  // Young account by age
  if (firstTradeTs) {
    try {
      const tradeDate = parseIsoOrUnix(tradeTimestamp);
      const accountAgeDays = (tradeDate.getTime() / 1000 - firstTradeTs) / 86400;

      if (accountAgeDays < 7) {
        boost = Math.max(boost, 2.0);
        signals.push({
          type: 'new_account',
          description: `New account: first trade was ${accountAgeDays.toFixed(1)} days ago`,
          value: accountAgeDays,
        });
      } else if (accountAgeDays < 30) {
        boost = Math.max(boost, 1.3);
        signals.push({
          type: 'young_account',
          description: `Young account: first trade was ${accountAgeDays.toFixed(0)} days ago`,
          value: accountAgeDays,
        });
      }
    } catch {
      // Invalid date, ignore
    }
  }

  return boost;
}

/**
 * Calculate standard deviation
 */
function calculateStdDev(values: number[], mean: number): number {
  if (values.length <= 1) return 0;

  const squaredDiffs = values.map((v) => Math.pow(v - mean, 2));
  const variance = squaredDiffs.reduce((a, b) => a + b, 0) / (values.length - 1);
  return Math.sqrt(variance);
}

/**
 * Parse ISO 8601 or Unix timestamp string to Date
 */
function parseIsoOrUnix(s: string): Date {
  const trimmed = s.trim();

  // Handle Unix timestamps (seconds or milliseconds)
  if (/^\d+$/.test(trimmed)) {
    let ts = parseInt(trimmed, 10);
    if (ts > 1e12) {
      ts = Math.floor(ts / 1000);
    }
    return new Date(ts * 1000);
  }

  // Handle ISO strings
  const isoStr = trimmed.replace('Z', '+00:00');
  return new Date(isoStr);
}
